src/v/cluster/rm_stm.cc

```diff
@@ -0,0 +1,1008 @@
+            _log_state.seq_table.emplace(bid.pid, entry);
+            _oldest_session = std::min(
+              _oldest_session, model::timestamp(entry.last_write_timestamp));
+        } else if (pid_seq->second.seq < bid.last_seq) {
+            pid_seq->second.seq = bid.last_seq;
+            pid_seq->second.last_write_timestamp = bid.max_timestamp.value();
+            _oldest_session = std::min(_oldest_session, bid.max_timestamp);
+        }
+    }
+
+    if (bid.is_transactional) {
+        if (_log_state.aborted.contains(bid.pid)) {
+            vlog(
+              clusterlog.error,
+              "Adding a record with pid:{} to a tx after it was aborted",
+              bid.pid);
+            if (_recovery_policy != best_effort) {
+                vassert(
+                  false,
+                  "Adding a record with pid:{} to a tx after it was aborted",
+                  bid.pid);
+            }
+            return;
+        }
+
+        if (_log_state.prepared.contains(bid.pid)) {
+            vlog(
+              clusterlog.error,
+              "Adding a record with pid:{} to a tx after it was prepared",
+              bid.pid);
+            if (_recovery_policy != best_effort) {
+                vassert(
+                  false,
+                  "Adding a record with pid:{} to a tx after it was prepared",
+                  bid.pid);
+            }
+            return;
+        }
+
+        auto ongoing_it = _log_state.ongoing_map.find(bid.pid);
+        if (ongoing_it != _log_state.ongoing_map.end()) {
+            if (ongoing_it->second.last < last_offset) {
+                ongoing_it->second.last = last_offset;
+            }
+        } else {
+            auto base_offset = model::offset(
+              last_offset() - (bid.record_count - 1));
+            _log_state.ongoing_map.emplace(
+              bid.pid,
+              tx_range{
+                .pid = bid.pid,
+                .first = base_offset,
+                .last = model::offset(last_offset)});
+            _log_state.ongoing_set.insert(base_offset);
+            _mem_state.estimated.erase(bid.pid);
+        }
+    }
+}
+
+void rm_stm::load_snapshot(stm_snapshot_header hdr, iobuf&& tx_ss_buf) {
+    vassert(
+      hdr.version == tx_snapshot_version,
+      "unsupported seq_snapshot_header version {}",
+      hdr.version);
+    iobuf_parser data_parser(std::move(tx_ss_buf));
+    auto data = reflection::adl<tx_snapshot>{}.from(data_parser);
+
+    for (auto& entry : data.fenced) {
+        _log_state.fence_pid_epoch.emplace(id(entry), epoch(entry));
+    }
+    for (auto& entry : data.ongoing) {
+        _log_state.ongoing_map.emplace(entry.pid, entry);
+        _log_state.ongoing_set.insert(entry.first);
+    }
+    for (auto& entry : data.prepared) {
+        _log_state.prepared.emplace(entry.pid, entry);
+    }
+    for (auto& entry : data.aborted) {
+        _log_state.aborted.emplace(entry.pid, entry);
+    }
```

Comment:
Maybe some range assignment would be more efficient. Are there any exception safety concerns?